cmds/mkdir/mkdir.go

```diff
@@ -16,32 +16,93 @@ package main
 import (
 	"flag"
 	"fmt"
+	"log"
 	"os"
+	"strconv"
+	"syscall"
+)
+
+type modeFlag struct {
+	set   bool
+	value string
+}
+
+func (m *modeFlag) String() string {
+	return m.value
+}
+
+func (m *modeFlag) Set(s string) error {
+	m.value = s
+	m.set = true
+	return nil
+}
+
+const (
+	cmd       = "mkdir [-m mode] [-v] [-p] <directory> [more directories]"
+	StickyBit = 01000
+	SgidBit   = 02000
+	SuidBit   = 04000
 )
 
 var (
+	mode    = &modeFlag{set: false, value: "0777"}
 	mkall   = flag.Bool("p", false, "Make all needed directories in the path")
-	mode    = flag.Int("m", 0666, "Directory mode")
 	verbose = flag.Bool("v", false, "Print each directory as it is made")
 	f       = os.Mkdir
 )
 
+func init() {
+	// Usage Definition
+	defUsage := flag.Usage
+	flag.Usage = func() {
+		os.Args[0] = cmd
+		defUsage()
+	}
+
+	// Complete Setting the flags
+	flag.Var(mode, "m", "Directory mode")
+}
+
 func main() {
 	flag.Parse()
 	if len(flag.Args()) < 1 {
-		fmt.Printf("Usage: mkdir [-m mode] [-v] [-p] <directory> [more directories]\n")
+		flag.Usage()
 		os.Exit(1)
 	}
 	if *mkall {
 		f = os.MkdirAll
 	}
+
+	// Get Correct Access Mode
+	if mode.set {
+		syscall.Umask(0)
+	}
+	accMode64bit, err := strconv.ParseUint(mode.value, 8, 32)
+	if err != nil || accMode64bit > 07777 {
+		log.Fatalf("invalid mode '%s'", mode.value)
+	}
+	accMode := os.FileMode(accMode64bit)
+	if accMode64bit&StickyBit != 0 {
+		accMode |= os.ModeSticky
+	}
+	if accMode64bit&SgidBit != 0 {
+		accMode |= os.ModeSetgid
+	}
+	if accMode64bit&SuidBit != 0 {
+		accMode |= os.ModeSetuid
+	}
+
 	for _, name := range flag.Args() {
-		if err := f(name, os.FileMode(*mode)); err != nil {
-			fmt.Printf("%v: %v\n", name, err)
+		if err := f(name, accMode); err != nil {
+			log.Fatalf("%v: %v\n", name, err)
 		} else {
 			if *verbose {
 				fmt.Printf("%v\n", name)
 			}
+			// os.Mkdir does not set up SGID and SUID correctly
+			if accMode64bit&SgidBit != 0 || accMode64bit&SuidBit != 0 {
```

Comment:
is  that a bug? this would be worth figuring out

Reply:
I have used strace to trace it down to the syscall level, the os.Mkdir calls the mkdirat syscall with the correct permission bits we passed in at that level. However, the SGID and SUID bits were still off. I think that might be a feature of the kernel.

Also, I just noticed that in the original mkdir command, whenever -m flag is used, the command makes a mkdir syscall with the permission bits we passed &~ the umask, following by a fchmod syscall with the actual permission bits we passed in.

So I think the fact that SGID and SUID not set up correctly in os.Mkdir is an expected behavior. Also, I have changed my code to reflect the original design of mkdir